### PR TITLE
[tests][s]: add tests for new actions - #32

### PR DESCRIPTION
**Pull request by mariorodeghiero**
Create tests to cover the actions `resource_sample_show` and `resource_schema_show` after the PR https://github.com/datopian/ckanext-blob-storage/pull/30

- [x] Add test for  `resource_sample_show`
- [x] Add test for  `resource_schema_show`

### CODE REVIEW

File: ckanext/external_storage/tests/test_actions.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
+from ckan.tests import factories
+from ckan.tests import helpers as test_helpers
+
+from ckanext.external_storage.tests import get_context
+
+
+class TestBlobSotrageActions():
```
Comment (shevron) [inline ckanext/external_storage/tests/test_actions.py:7]: With `pytest`, there is really almost never a good reason to put tests in a class. With modern fixtures, this is typically just an overhead that is not required and may cause issues

File: ckanext/external_storage/tests/test_actions.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    """Test cases for logic actions
+    """
+
+    org_admin = factories.User()
```
Comment (shevron) [inline ckanext/external_storage/tests/test_actions.py:11]: These class-level objects (org_admin, initial_dataset, etc.) are not good, because they are *static* - this is state that is not cleared between tests. Better to use fixtures, or if we have to, test setup / tear-down functions to create them and clear them after each test.

File: ckanext/external_storage/tests/test_actions.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        name="First Resource",
+        package_id=initial_dataset['id'],
+        sample={'1': {'VMP_SNOMED_CODE': '4747111000001104'}},
+        schema={'fields': [{'type': 'string'}], 'missingValues': ['']}
+    )
+
+    def test_resource_schema_show(self):
+        """Test if the resource.schema is a string the resource_sample_show should return JSON object
+        """
+        context = get_context(self.org_admin)
```
Comment (shevron) [inline ckanext/external_storage/tests/test_actions.py:31]: You don't need `get_context` (see above) here. It would be cleaner to use the existing `user_context(user)`  context manager:

```python
def test_resource_schema_show(self):
    with user_context(self.org_admin) as context:
        resource_schema = test_helpers.call_action(
            'resource_schema_show',
            context,
            id=self.first_resource['id'])
    expect = {'fields': [{'type': 'string'}], 'missingValues': ['']}
    assert resource_schema == expect
```

BTW, context managers are an awesome Python feature. You can read about them more for example here: https://towardsdatascience.com/the-magic-of-python-context-managers-adb92ace1dd0. Understanding how they work and when to use them is a very useful tool for Python developers.

File: ckanext/external_storage/tests/test_actions.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        expect = {'1': {'VMP_SNOMED_CODE': '4747111000001104', 'ODS_CODE': 'REF'}}
+
+        resource_sample = test_helpers.call_action(
+            'resource_sample_show',
+            context,
+            id=self.first_resource['id'])
+
+        assert resource_sample == expect
+
+    def test_resource_schema_show_and_resource_sample_show_with_new_ckan_version(self):
```
Comment (shevron) [inline ckanext/external_storage/tests/test_actions.py:54]: How is this different from the other two tests? It tests the exact same thing(s). Suggest to just delete it and make sure the test suite runs on both CKAN 2.8.4 and 2.8.5 in CI.

File: ckanext/external_storage/tests/__init__.py [hunk 5]
```diff
@@ -11,15 +11,25 @@ class FunctionalTestBase(helpers.FunctionalTestBase):
...
 @contextmanager
 def user_context(user):
     # type: (Dict[str, Any]) -> Dict[str, Any]
     """Context manager that creates a CKAN context dict for a user, then
     both patches our `get_user_context` function to return it and also
     yields the context for use inside tests
     """
-    userobj = model.User.get(user['name'])
-    context = {"model": model,
+    userobj = core_model.User.get(user['name'])
```
Comment (shevron) [inline ckanext/external_storage/tests/__init__.py:31]: DRY. If you wrote `get_context`, factor out the identical code from here and just call `get_context`. But really, I'm not sure you need this, because you can probably just use `user_context(user)` wherever you need `get_context()` in your tests.

**Review by shevron — CHANGES_REQUESTED** [review]
This is a great start but I would like to improve it a bit so it can serve as a basis for more action tests. If the feedback I gave is a bit too much, we can either schedule a pair programming session this week or I can just fix it myself and you will review.

**Comment by rufuspollock** [comment]
@mariorodeghiero i'm not quite sure why we are adding resource_schema_show and resource_sample_show actions and hence why we have this PR. Could you give more detail on why we are adding thos actions in #30.

**Comment by mariorodeghiero** [comment]
@rufuspollock I think you check more details in https://github.com/datopian/ckanext-blob-storage/issues/29 if is not clear I suggest you talk to @anuveyatsu because propose the solution because is too hard to parse it on the front-end side.

**Comment by shevron** [comment]
@mariorodeghiero I'm closing this as it became stale. If you ever want to refresh this and re-open, feel free to do so.

**Comment by mariorodeghiero** [comment]
@shevron thanks :+1:
